### PR TITLE
Feature rcpp expectation

### PR DESCRIPTION
**Pull request by niels89**
Opened pull request, tests and documentation are still missing.

### CODE REVIEW

File: src/bgnbd_expectation.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
+#include <RcppArmadillo.h>
+#include <math.h>
+#include "bgnbd_expectation.h"
+#include "clv_vectorized.h"
+
+//' @name bgnbd_expectation
+//' @title BG/NBD: Unconditional Expectation
+//'
+//' @description Computes the expected number of repeat transactions in the interval (0, vT_i]
+//' for a randomly selected customer, where 0 is defined as the point when the customer came alive.
```
Comment (pschil) [inline src/bgnbd_expectation.cpp:10]: move to template and re-use in all rcpp expectation functions (bgnbd, pnbd, ggomnd)

File: src/bgnbd_expectation.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+//'
+//' @template template_params_bgnbd
+//' @param vT_i Number of periods since the customer came alive
```
Comment (pschil) [inline src/bgnbd_expectation.cpp:13]: move @param vT_i docu to template, re-use in all expectation functions, incl ggomnbd

File: src/bgnbd_expectation.cpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+//' @template template_params_rcppcovmatrix
+//' @template template_params_rcppvcovparams
+//'
+//' @templateVar name_params_cov_life vCovParams_life
+//' @templateVar name_params_cov_trans vCovParams_trans
+//' @template template_details_rcppcovmatrix
+//'
+//' @template template_references_bgnbd
```
Comment (pschil) [inline src/bgnbd_expectation.cpp:21]: return value needs to be documented as well. unfortunately also not done in ggomnd expectation... make template, re-use across expectation functions

File: src/bgnbd_expectation.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+                                     const double b,
+                                     const arma::vec& vT_i,
+                                     const arma::vec& vCovParams_trans,
+                                     const arma::vec& vCovParams_life,
+                                     const arma::mat& mCov_life,
+                                     const arma::mat& mCov_trans){
+
+  const arma::vec vAlpha_i = alpha * arma::exp(((mCov_trans * (-1)) * vCovParams_trans));
+  const arma::vec vA_i  = a * arma::exp(mCov_life * vCovParams_life);
+  const arma::vec vB_i  = b * arma::exp(mCov_life * vCovParams_life);
```
Comment (pschil) [inline src/bgnbd_expectation.cpp:75]: the calculations of alpha_i, beta_i, etc are done so often across all methods and files and should (finally..) be put in separate methods, for all models

File: src/pnbd_expectation.cpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  // Build alpha and beta -------------------------------------------
+  //    With static covariates: alpha and beta different per customer
+  //
+  //    alpha_i: alpha0 * exp(-cov.trans * cov.params.trans)
+  //    beta_i:  beta0  * exp(-cov.life  * cov.parama.life)
+  const arma::vec vAlpha_i = alpha_0 * arma::exp(((mCov_trans * (-1)) * vCovParams_trans));
+  const arma::vec vBeta_i  = beta_0  * arma::exp(((mCov_life  * (-1)) * vCovParams_life));
+  arma::vec vR(vAlpha_i.n_elem);
+  vR.fill(r);
```
Comment (pschil) [inline src/pnbd_expectation.cpp:73]: vR obsolete

File: tests/testthat/test_correctness_bgnbd.R [hunk 6]
```diff
@@ -20,4 +20,112 @@ fct.testthat.correctness.clvfitted.correct.coefs(method = bgnbd,
...
+                                       date.format = "ymd", time.unit = "W", estimation.split = 38,
+                                       name.id = "Id", name.date = "Date", name.price = "Price"))
+
+  clv.apparel.static <- SetStaticCovariates(clv.data = clv.apparel,
+                                            data.cov.life = apparelStaticCov,
+                                            data.cov.trans = apparelStaticCov,
+                                            names.cov.life = "Gender",
+                                            names.cov.trans = "Gender",
+                                            name.id = "Id")
+
```
Comment (pschil) [inline tests/testthat/test_correctness_bgnbd.R:85]: use function `fct.helper.create.clvdata.apparel.staticcov` to create the object

File: tests/testthat/test_correctness_pnbd.R [hunk 7]
```diff
@@ -49,4 +49,102 @@ fct.testthat.correctness.dyncov(data.apparelTrans=apparelTrans, data.apparelDynC
...
+  # No cov ---------------------------------------------------------------------------------------------------
+  expect_silent(clv.cdnow <- clvdata(data.transactions = cdnow,
+                                     date.format = "ymd", time.unit = "W", estimation.split = 38,
+                                     name.id = "Id", name.date = "Date", name.price = "Price"))
+
+
+  expect_silent(obj.fitted <- do.call(pnbd, list(clv.data = clv.cdnow, verbose = FALSE)))
+
+
+  params_i <- obj.fitted@cbs[, c("Id", "T.cal", "date.first.actual.trans")]
```
Comment (pschil) [inline tests/testthat/test_correctness_pnbd.R:65]: This seems very repetitive, same for pnbd and bgnbd and nearly same for cov and nocov. Could that not be put in a separate function in one of the "helper_*" files (has to be in "helper_" file) and then re-used?

File: tests/testthat/test_correctness_pnbd.R [hunk 8]
```diff
@@ -49,4 +49,102 @@ fct.testthat.correctness.dyncov(data.apparelTrans=apparelTrans, data.apparelDynC
...
+                                       date.format = "ymd", time.unit = "W", estimation.split = 38,
+                                       name.id = "Id", name.date = "Date", name.price = "Price"))
+
+  clv.apparel.static <- SetStaticCovariates(clv.data = clv.apparel,
+                                            data.cov.life = apparelStaticCov,
+                                            data.cov.trans = apparelStaticCov,
+                                            names.cov.life = "Gender",
+                                            names.cov.trans = "Gender",
+                                            name.id = "Id")
+
```
Comment (pschil) [inline tests/testthat/test_correctness_pnbd.R:108]: use `fct.helper.create.clvdata.apparel.staticcov`

**Comment by niels89** [comment]
>     * [ ]  consistent cpp: rename all .h to .hpp, also other existing ones
> 
Seems like Rcmd checks don't like the suffix .hpp:

checking if this is a source package ... WARNING
Subdirectory 'src' contains:
    bgnbd_expectation.hpp clv_vectorized.hpp ggomnbd_LL.hpp
    ggomnbd_PAlive.hpp ggomnbd_expectation.hpp pnbd_LL_ind.hpp
    pnbd_PAlive.hpp pnbd_expectation.hpp
  These are unlikely file names for src files.

**Comment by niels89** [comment]
@pschil  I have fixed all review issues and updated the UML diagrams to account for the latest changes.

File: src/pnbd_expectation.cpp [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+arma::vec pnbd_nocov_expectation(const double r,
+                                    const double s,
+                                    const double alpha_0,
+                                    const double beta_0,
+                                    const arma::vec& vT_i){
+
+  // Build alpha and beta --------------------------------------------------------
+  const double n = vT_i.n_elem;
+  arma::vec vAlpha_i(n), vBeta_i(n);
+
```
Comment (pschil) [inline src/pnbd_expectation.cpp:44]: dont alloc mem, assign return from pnbd_nocov_i

File: src/pnbd_expectation.cpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+                                        const arma::vec& vCovParams_trans,
+                                        const arma::vec& vCovParams_life,
+                                        const arma::mat& mCov_life,
+                                        const arma::mat& mCov_trans){
+
+  // Build alpha and beta -------------------------------------------
+  //    With static covariates: alpha and beta different per customer
+  //
+  //    alpha_i: alpha0 * exp(-cov.trans * cov.params.trans)
+  //    beta_i:  beta0  * exp(-cov.life  * cov.parama.life)
```
Comment (pschil) [inline src/pnbd_expectation.cpp:71]: remove comments

File: tests/testthat/test_correctness_pnbd.R [hunk 11]
```diff
@@ -49,4 +49,76 @@ fct.testthat.correctness.dyncov(data.apparelTrans=apparelTrans, data.apparelDynC
...
+                                                                                               s = obj.fitted@prediction.params.model[["s"]],
+                                                                                               alpha_0 = obj.fitted@prediction.params.model[["alpha"]],
+                                                                                               beta_0 = obj.fitted@prediction.params.model[["beta"]],
+                                                                                               vT_i = t_i,
+                                                                                               vCovParams_trans = obj.fitted@prediction.params.trans,
+                                                                                               vCovParams_life = obj.fitted@prediction.params.life,
+                                                                                               mCov_life = m.cov.data.trans,
+                                                                                               mCov_trans = m.cov.data.life)])}
```
Comment (pschil) [inline tests/testthat/test_correctness_pnbd.R:118]: trans vs life mixed up

File: tests/testthat/helper_testthat_correctness_transactions.R [hunk 12]
```diff
@@ -231,3 +231,37 @@ fct.testthat.correctness.clvfittedtransactions <- function(name.model, method, d
   fct.testthat.correctness.clvfittedtransactions.staticcov.regularization.lambda.0.no.regularization(method = method, clv.apparel.staticcov = clv.apparel.staticcov,
                                                                                                      m.fitted.static = obj.fitted.static)
 }
+
+fct.testthat.correctness.clvfittedtransactions.common <- function(fct.expectation.R, fct.expectation.Cpp, params_i, obj.fitted){
+  dt.expectation.seq <- clv.time.expectation.periods(clv.time = obj.fitted@clv.data@clv.time,
```
Comment (pschil) [inline tests/testthat/helper_testthat_correctness_transactions.R:236]: name method more descriptive. Something like`fct.testthat.correctness.clvfittedtransactions.same.expectation.in.R.and.Cpp`

File: tests/testthat/helper_testthat_correctness_transactions.R [hunk 13]
```diff
@@ -231,3 +231,37 @@ fct.testthat.correctness.clvfittedtransactions <- function(name.model, method, d
...
+
+fct.testthat.correctness.clvfittedtransactions.staticcov <- function(method, apparelTrans, apparelStaticCov){
+  clv.apparel.static <- fct.helper.create.clvdata.apparel.staticcov(data.apparelTrans = apparelTrans,
+                                                                    data.apparelStaticCov = apparelStaticCov,
+                                                                    estimation.split = 38)
+
+  expect_silent(obj.fitted <- do.call(method, list(clv.data = clv.apparel.static, verbose = FALSE)))
+
+  return(obj.fitted)
+}
```
Comment (pschil) [inline tests/testthat/helper_testthat_correctness_transactions.R:267]: dont put fitting the model on data into separate functions. would replace at dozens of places where now the method is called inplace (ie `pnbd(clv.cdnow`). Rather move the code here to the "Expectation in Rcpp matches expecttation ... " tests

File: tests/testthat/test_correctness_bgnbd.R [hunk 14]
```diff
@@ -20,4 +20,85 @@ fct.testthat.correctness.clvfitted.correct.coefs(method = bgnbd,
                                                  params.nocov.coef = c(r = 0.243, alpha = 4.414, a = 0.793, b = 2.426),
                                                  LL.nocov = -9582.4)
 
+context("Correctness - BG/NBD nocov - Expectation")
 
+test_that("Expectation in Rcpp matches expectation in R (nocov)", {
+
+  # No cov ---------------------------------------------------------------------------------------------------
```
Comment (pschil) [inline tests/testthat/test_correctness_bgnbd.R:27]: mark with `skip_on_cran()`

File: tests/testthat/test_correctness_bgnbd.R [hunk 15]
```diff
@@ -20,4 +20,85 @@ fct.testthat.correctness.clvfitted.correct.coefs(method = bgnbd,
...
+                                                        fct.expectation.R = fct.expectation.R,
+                                                        params_i = params_i,
+                                                        obj.fitted = obj.fitted)
+})
+
+context("Correctness - BG/NBD staticcov - Expectation")
+
+test_that("Expectation in Rcpp matches expectation in R (staticcov)", {
+
+  # Static cov ---------------------------------------------------------------------------------------------------
```
Comment (pschil) [inline tests/testthat/test_correctness_bgnbd.R:62]: `skip_on_cran()`

File: tests/testthat/test_correctness_pnbd.R [hunk 16]
```diff
@@ -49,4 +49,76 @@ fct.testthat.correctness.dyncov(data.apparelTrans=apparelTrans, data.apparelDynC
 
 # ** Static cov??
 
+context("Correctness - PNBD nocov - Expectation")
 
+test_that("Expectation in Rcpp matches expectation in R (nocov)", {
+
+  # No cov ---------------------------------------------------------------------------------------------------
```
Comment (pschil) [inline tests/testthat/test_correctness_pnbd.R:56]: `skip_on_cran()`

File: tests/testthat/test_correctness_pnbd.R [hunk 17]
```diff
@@ -49,4 +49,76 @@ fct.testthat.correctness.dyncov(data.apparelTrans=apparelTrans, data.apparelDynC
...
+                                                        params_i = params_i,
+                                                        obj.fitted = obj.fitted)
+
+})
+
+context("Correctness - PNBD staticcov - Expectation")
+
+test_that("Expectation in Rcpp matches expectation in R (staticcov)", {
+
+  # Static cov ---------------------------------------------------------------------------------------------------
```
Comment (pschil) [inline tests/testthat/test_correctness_pnbd.R:87]: `skip_on_cran()`

File: tests/testthat/helper_testthat_correctness_transactions.R (hunk 13 above)
Comment (niels89) [reply tests/testthat/helper_testthat_correctness_transactions.R:267]: I think I don't understand what I should change here. Would you remove the clvfittedtransactions.nocov / staticcov altogether and put the whole code inside fct.testthat.correctness.clvfittedtransactions.same.expectation.in.R.and.Cpp?

**Pull request merged**